src/controllers/plugins/PluginController.cpp

```diff
@@ -0,0 +1,286 @@
+
+    lua_pushnil(L);
+    lua_setfield(L, gtable, "loadfile");
+
+    lua_pushnil(L);
+    lua_setfield(L, gtable, "dofile");
+
+    lua_pop(L, 1);
+}
+
+void PluginController::load(const QFileInfo &index, const QDir &pluginDir,
+                            const PluginMeta &meta)
+{
+    lua_State *l = luaL_newstate();
+    PluginController::openLibrariesFor(l, meta);
+
+    auto pluginName = pluginDir.dirName();
+    auto plugin = std::make_unique<Plugin>(pluginName, l, meta, pluginDir);
+    this->plugins_.insert({pluginName, std::move(plugin)});
+    if (!PluginController::isEnabled(pluginName))
+    {
+        qCInfo(chatterinoLua) << "Skipping loading" << pluginName << "("
+                              << meta.name << ") because it is disabled";
+        return;
+    }
+    qCDebug(chatterinoLua) << "Running lua file:" << index;
+    int err = luaL_dofile(l, index.absoluteFilePath().toStdString().c_str());
+    if (err != 0)
+    {
+        qCWarning(chatterinoLua)
+            << "Failed to load" << pluginName << "plugin from" << index << ": "
+            << lua::humanErrorText(l, err);
+        return;
+    }
+    qCInfo(chatterinoLua) << "Loaded" << pluginName << "plugin from" << index;
+}
+
+bool PluginController::reload(const QString &id)
+{
+    auto it = this->plugins_.find(id);
+    if (it == this->plugins_.end())
+    {
+        return false;
+    }
+    if (it->second->state_ != nullptr)
+    {
+        lua_close(it->second->state_);
+        it->second->state_ = nullptr;
+    }
+    for (const auto &[cmd, _] : it->second->ownedCommands)
+    {
+        getApp()->commands->unregisterPluginCommand(cmd);
+    }
+    it->second->ownedCommands.clear();
+    QDir loadDir = it->second->loadDirectory_;
+    this->plugins_.erase(id);
+    this->tryLoadFromDir(loadDir);
+    return true;
+}
+
+QString PluginController::tryExecPluginCommand(const QString &commandName,
+                                               const CommandContext &ctx)
+{
+    for (auto &[name, plugin] : this->plugins_)
+    {
+        if (auto it = plugin->ownedCommands.find(commandName);
+            it != plugin->ownedCommands.end())
+        {
+            const auto &funcName = it->second;
+
+            auto *L = plugin->state_;
+            lua_getfield(L, LUA_REGISTRYINDEX, funcName.toStdString().c_str());
+            lua::push(L, ctx);
+
+            auto res = lua_pcall(L, 1, 0, 0);
+            if (res != LUA_OK)
+            {
+                ctx.channel->addMessage(makeSystemMessage(
+                    "Lua error: " + lua::humanErrorText(L, res)));
+                return "";
+            }
+            return "";
+        }
+    }
+    qCCritical(chatterinoLua)
+        << "Something's seriously up, no plugin owns command" << commandName
+        << "yet a call to execute it came in";
+    assert(false && "missing plugin command owner");
+    return "";
+}
+
+bool PluginController::isEnabled(const QString &id)
```

Comment:
Maybe `isPluginEnabled` instead of just `isEnabled`?

Also might want to not do the `pluginSupportEnabled` check here, but rather leave that up to the caller instead